Provide semaphores in the threading library

xavierleroy opened the pull request:
This PR proposes to add a new thread-related module Semaphore, implementing [Dijkstra's semaphores](https://en.wikipedia.org/wiki/Semaphore_(programming)).

Semaphores would be useful for code that currently and incorrectly uses mutexes as binary semaphores, see discussion at #9757 , #9846, and especially https://github.com/ocaml/ocaml/pull/9846#issuecomment-691672306 .

xavierleroy committed:
Provide semaphores

This commit adds a new thread-related module Semaphore, implementing
counting semaphores and binary semaphores.

dbuenzli commented on otherlibs/systhreads/semaphore.ml:
```diff
@@ -0,0 +1,67 @@
+  if v < 0 || v > m then
+    raise (Sys_error "Semaphore.init: wrong initial value");
+  if m <= 0 then
+    raise (Sys_error "Semaphore.init: wrong maximal value");
```
It would be nice if these two error messages mentioned the offending values and how they offend.

dbuenzli commented on otherlibs/systhreads/semaphore.ml:
```diff
@@ -0,0 +1,67 @@
+    raise (Sys_error "Semaphore.init: wrong maximal value");
+  { mut = Mutex.create(); v; max; nonzero = Condition.create() }
+
+let make_binary b =
```
Why not use `make` directly ?

xavierleroy replied:
Just to avoid the checks on the init value and the max value, which are trivially true here.

dbuenzli commented on otherlibs/systhreads/semaphore.ml:
```diff
@@ -0,0 +1,67 @@
+        if s.v < max_int then
+          s.v <- s.v + 1
+        else begin
+          Mutex.unlock s.mut;
```
We'll only be in this branch if `s.v = max_int` I don't mind it that way but I think using this as the test would have been clearer. (Also somehow you could have avoided the option for `max` altogether but maybe it's clearer that way).

xavierleroy replied:
The code is written like this for symmetry between the `None` and the `Some` cases.  "if current value less than threshold, increment value, otherwise ..."

dbuenzli commented on otherlibs/systhreads/semaphore.mli:
```diff
@@ -0,0 +1,80 @@
+type t
+(** The type of semaphores. *)
+
+val make : ?max: int -> int -> t
```
I have the impression this is going to be a lost cause :-) but anyway.

So I `create` my threads, my condition variables and my mutexes. However I `make` my atomic variables and my semaphores ? I'm afraid [this convention](https://github.com/ocaml/ocaml/pull/9921#issuecomment-693465196) only make sense to people who are in the know. 

If your convention cannot be understood so as to raise questions on [stackoverflow](https://stackoverflow.com/questions/29363460/whats-the-ocaml-naming-convention-for-constructors) maybe you should ditch it (especially since it can be changed *now* without too much inconsistency churn).

xavierleroy replied:
It is a lost cause, I'm afraid :-)

dbuenzli commented on otherlibs/systhreads/semaphore.mli:
```diff
@@ -0,0 +1,80 @@
+    ([false] means [0], [true] means [1]) and maximal value [1].
+*)
+
+val post : t -> unit
```
Personally I don't like to use `post` when there's no transmitted payload. I would rather use `signal` here or even switch the whole abstraction to `acquire/release` which I find clearer for resource handling code which is what you end up doing with semaphores most of the time.

xavierleroy replied:
POSIX uses `sem_post`.

dbuenzli commented on otherlibs/systhreads/semaphore.mli:
```diff
@@ -0,0 +1,80 @@
+    is not zero, then atomically decrements the value of [s] and returns.
+*)
+
+val trywait : t -> bool
```
I would spell that `try_wait`. But I find the name a bit absurd since you never end-up waiting. If you switch the terminology to `acquire/release` the name would be `try_acquire` and would make more sense.

xavierleroy replied:
POSIX uses `sem_trywait`, probably for symmetry with `pthread_mutex_trylock`.  I agree it's a bit silly, though.

dbuenzli reviewed (commented):
I find this to be a useful addition to the stdlib. However: 

1. I would change the terminology towards `acquire/release` and use `capacity` for `max`
2. Provide accessors for `capacity` and the current value. 
3. Provide the ability to dynamically resize the capacity of a semaphore so that it can be used for dynamic rate throttling (that does change a few invariants though). 
4. Maybe provide brackets `with_resource` (or `with_permit` if you reuse e.g. the Java terminology)  that `acquire/release` a thunk and does the `Fun.protect` for you.

xavierleroy commented:
Thanks for the review.  A quick reply to the high-level points:

> I would change the terminology towards acquire/release and use capacity for max

The POSIX operation names are `sem_post`, `sem_wait`, `sem_trywait`, and the OCaml thread library tries to look like POSIX threads.  This said, I find acquire/release quite nice too.  Let's wait for some other opinions.

> Provide accessors for capacity and the current value.

Accessing the current value is inherently racy.  Let's not offer our users this opportunity to shoot themselves in the foot.  Without access to the current value, access  to the max value is not particularly useful either.

> Provide the ability to dynamically resize the capacity of a semaphore so that it can be used for dynamic rate throttling (that does change a few invariants though).

I didn't think of this.  Not sure I like the extra complexity.

> Maybe provide brackets with_resource (or with_permit if you reuse e.g. the Java terminology) that acquire/release a thunk and does the Fun.protect for you.

This would be a must-have for the Mutex module.  But many interesting uses of semaphores do not bracket P and V operations.  See the [producer-consumer example](https://en.wikipedia.org/wiki/Semaphore_(programming)#Producer%E2%80%93consumer_problem): semaphore operations on `fullCount` and `emptyCount` are not bracketed!  `useCount` is bracketed, but that's because it is used as a mutex.

dbuenzli commented:
> Accessing the current value is inherently racy.

It's useful however for reporting interfaces, when you use semaphores for ressource consumption throttling. 

> The POSIX operation names are `sem_post`, `sem_wait`, `sem_trywait`, and the OCaml thread library tries to look like POSIX threads. This said, I find acquire/release quite nice too. Let's wait for some other opinions.

Just as point of reference you may want to look into the terminology of [Java semaphores](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Semaphore.html) or Lucene's [AdjustableSemaphore](https://lucene.apache.org/solr/7_7_1/solr-core/org/apache/solr/util/AdjustableSemaphore.html).

Octachron commented:
Should the new module be added to the list in  `manual/manual/library/libthreads.etex` ?

xavierleroy committed:
Add the Semaphore module to the manual

xavierleroy commented:
> Should the new module be added to the list in manual/manual/library/libthreads.etex ?

It should!  See commit 9e91f9dce .

xavierleroy committed:
Semaphore: use "acquire" and "release" terminology

xavierleroy commented:
I tried the "release"/"acquire" terminology suggested by @dbuenzli .  (Commit 568aa66.)  Opinions welcome.

stedolan commented:
Because of `capacity`, these are not exactly Dijkstra's semaphores, and I'm not quite convinced by the current semantics..

Dijkstra (e.g. in [EWD123](https://www.cs.utexas.edu/users/EWD/transcriptions/EWD01xx/EWD123-2.html)) defines _general semaphores_ as the normal unbounded counters, implemented here when `capacity = None`. As far as I'm aware, his _binary semaphores_ are the same structure, except proven never to go above one. I don't think any of his work gives semantics to releasing a binary semaphore twice (My understanding is if that could occur, he wouldn't call that semaphore "binary" any more).

If we do add binary semaphores, restricted to values 0 and 1, I see three possible semantics for releasing one twice:

  1. Releasing a binary semaphore that is already at 1 is a no-op. I think this is the semantics of Win32 auto-reset Event objects (with acquire = `WaitForSingleObject` and release = `SetEvent`).

  2. Releasing a binary semaphore that is already at 1 is an error which raises an exception. This is the semantics of "mutexes that can be unlocked from another thread"

  3. Releasing a binary semaphore that is already at 1 blocks until the semaphore is zero, then sets it to one. This is the semantics of a unit `MVar` in Haskell, or of a bounded queue of size one.

It's not obvious to me that the current choice, (1), is the best one. I'm also unsure of the use-case of values of `capacity` between 1 and infinity. (Under semantics (3), this could model a bounded queue of tokens, but under semantics (1) I'm not sure what it's for)

Also @dbuenzli : an ordinary semaphore works fine for dynamic rate throttling, no `capacity` required. Each worker `acquire`s the semaphore when starting a task and `release`s it when done. The throttle uses `release` to increase the maximum number of parallel workers, and `acquire` to decrease it.

damiendoligez commented on otherlibs/systhreads/semaphore.ml:
```diff
@@ -0,0 +1,67 @@
+  let ret = if s.v = 0 then false else (s.v <- s.v - 1; true) in
+  Mutex.unlock s.mut;
+  ret
+
```
Let's make `check-typo` happy...

```suggestion
```

gadmm commented:
(Paging @CraigFe who might be able to comment on which semantics are expected for backwards compatibility.)

damiendoligez reviewed (commented):
@stedolan You can implement (3) with a pair of general semaphores, and you get dynamic resizing for free.

@xleroy
- ~~I don't see the point of bounded semaphores, and they don't seem to be in POSIX.~~
- I think `getvalue` would be useful for monitoring and debugging (and it *is* in POSIX).

craigfe commented:
Thanks @gadmm for the ping.

> 1. Releasing a binary semaphore that is already at 1 is a no-op. I think this is the semantics of Win32 auto-reset Event objects (with acquire = `WaitForSingleObject` and release = `SetEvent`).
> 
> 2. Releasing a binary semaphore that is already at 1 is an error which raises an exception. This is the semantics of "mutexes that can be unlocked from another thread"
>
>   3. Releasing a binary semaphore that is already at 1 blocks until the semaphore is zero, then sets it to one. This is the semantics of a unit `MVar` in Haskell, or of a bounded queue of size one.

Regarding our specific use-cases in Mirage for delegating resource ownership to child threads, any of the three options is fine.

On my machine at least, calling pre-4.12 `Mutex.unlock` on an unheld lock is a no-op, so behaves as (1) rather than (2). This perhaps makes (1) a better drop-in replacement for the old `Mutex`. Personally, I much prefer (2), but it is easy for me to say that knowing I don't rely on the old behaviour. The counter-argument here is that it's fairly easy to implement (1) in terms of (2).

I certainly wouldn't expect a "semaphore" to behave as (3), and remember being taught to implement bounded queues with a pair of semaphores as @damiendoligez suggests.

xavierleroy commented:
Thank you all for the feedback.

Concerning the semantics of binary semaphores, I fully agree with @CraigFe.  I think we want binary semaphores that act like non-error-checking, mistake-tolerant mutexes, namely: 

1. a mutex can be unlocked from a thread other than the one that locked it, and
2. unlocking an unlocked mutex is a no-op.  

That's how mutexes behaved in the VM-level implementation of OCaml threads, and I believe that's how default mutexes (without HLE) behave in Glibc NPTL.  For better or worth, some users have come to rely on this behavior.  So, if we tell them "you can't use Mutex like this any longer but you should use Semaphore instead", the semaphores should better have this behavior, corresponding to @stedolan 's semantics number 1.

Concerning the "capacity" API: this was one way to unify counted semaphores and binary semaphores, and share acquire and release operations between them.  I agree that a counted semaphore with limited capacity other than 1 is rather useless.  The alternative would be to have two APIs, one for counted semaphores and one for binary semaphores.  I'll sketch that in another message.

xavierleroy commented:
Here is an alternate API, with two sub-modules for the `Semaphore` compilation unit:
```
module Counted : sig
  type t
  val make: int -> t
  val release: t -> unit  (* possibly: an optional argument of type int to release several units at once *)
  val acquire: t -> unit  (* likewise *)
  val try_acquire: t -> bool (* likewise *)
  val current_value: t -> int   (* to please @damiendoligez and @dbuenzli *)
end

module Binary : sig
  type t
  val make: bool -> t
  val release: t -> unit             (* set to "true" state, whether the previous state was "false" or "true" *)
  val acquire: t -> unit             (* wait until in the "true" state and set to "false" *)
  val try_acquire: t -> bool
  (*  no "current_value" because we don't have that for mutexes either? *)
end

```

stedolan commented:
Having separate submodules for Counted and Binary makes sense to me.

Separating the submodules means we need not give the same names to the operations of counted and binary semaphores. If the goal is a drop-in replacement for the default-NPTL-mutex behaviour, should we call the operations in Binary `lock` and `unlock`? That way, users who rely on this behaviour can write `module Mutex = Semaphore.Binary` at the top of their files, and won't need to change all their code.

craigfe commented:
Personally, I find the term `release` helpful in differentiating the behaviour of mutexes vs. binary semaphores. For those users who will need to make the switch, the difference between the new `Mutex.unlock` and `Semaphore.Binary.release` is presumably relevant to the correctness of their code. Thus it seems worthwhile to have an API that makes this distinction clear and does not encourage patterns like `module Mutex = Semaphore.Binary` that obfuscate it for future readers of that code.

I really like the two submodules proposal. To bikeshed, I suggest `Counting` rather than `Counted`; the semaphores are the things doing the counting, they are not counted themselves.

xavierleroy commented:
>  If the goal is a drop-in replacement for the default-NPTL-mutex behaviour, 

"Drop-in" to some extent only: these binary semaphores can't be used with condition variables (the `Condition.wait` operation), and taking `module Mutex = Semaphore.Binary` will break as soon as condition variables are involved.

xavierleroy committed:
Add the Semaphore module to the manual, continued

xavierleroy committed:
New API for the Semaphore module

Counting semaphores and binary semaphores are presented as two different
abstract types in two different sub-modules.
Counting semaphores no longer have a optional maximal capacity.

xavierleroy commented:
The new API, with two sub-modules, is in commit 437adec .

damiendoligez commented on otherlibs/systhreads/semaphore.ml:
```diff
@@ -0,0 +1,87 @@
+
+let make v =
+  if v < 0 then
+    raise (Sys_error "Semaphore.Counting.init: wrong initial value");
```
Shouldn't this raise `Invalid_argument` rather than `Sys_error`?

xavierleroy replied:
You're right, will change.

damiendoligez commented on otherlibs/systhreads/semaphore.ml:
```diff
@@ -0,0 +1,87 @@
+    Mutex.unlock s.mut
+  end else begin
+    Mutex.unlock s.mut;
+   raise (Sys_error "Semaphore.Counting.release: overflow")
```
```suggestion
    raise (Sys_error "Semaphore.Counting.release: overflow")
```

damiendoligez commented on otherlibs/systhreads/semaphore.mli:
```diff
@@ -0,0 +1,140 @@
+(** {2 Binary semaphores} *)
+
+(** Binary semaphores are a variant of counting semaphores
+    where semaphores can only take two values, 0 and 1.
```
Since the API uses only booleans, I think you should document it in terms of booleans, not integers.
```suggestion
    where semaphores can only take two values, [false] and [true].
```

xavierleroy replied:
I tried but it makes it harder to see why it's a special case of counting semaphores.  I think it reads better with zeroes and ones.

damiendoligez reviewed (approved):
A few suggestions, but this is good to merge already.

xavierleroy committed:
Wrong indentation

Co-authored-by: Damien Doligez <damien.doligez@gmail.com>

xavierleroy committed:
Raise Invalid_argument instead of Sys_error for wrong initial value

xavierleroy committed:
Changes for 9930
